exercises/simple-linked-list/example.js

```diff
@@ -0,0 +1,115 @@
+export class Element {
+  constructor(value, next) {
+    if (value === undefined) {
+      throw new Error('Value required.');
+    }
+
+    if (next !== undefined && !(next instanceof Element)) {
+      throw new Error('A Element instance as next value is required.');
+    }
+
+    this.value = value;
+    this.next = next;
+  }
+}
+
+export class List {
+  static fromArray(array) {
+    const list = new List();
+    array.forEach(item => list.push(new Element(item)));
+
+    return list;
+  }
+
+  push(value) {
+    if (value === undefined) {
+      throw new Error('Argument required.');
+    }
+
+    const newEl = (value instanceof Element)
+      ? value
+      : new Element(value);
+
+    if (!this.head) {
+      this.head = newEl;
+      return;
+    }
+
+    let lastEl = this.head;
+    while (lastEl.next) {
+      lastEl = lastEl.next;
+    }
+
+    lastEl.next = newEl;
+  }
+
+  unshift(value) {
+    if (value === undefined) {
+      throw new Error('Argument required.');
+    }
+
+    const newEl = (value instanceof Element)
+      ? value
+      : new Element(value);
+
+    newEl.next = this.head;
+    this.head = newEl;
+  }
+
+  shift() {
+    if (!this.head) {
+      return;
+    }
+
+    this.head = this.head.next;
+  }
+
+  pop() {
+    if (!this.head) {
+      return;
+    }
+
+    let penultEl;
+    let lastEl = this.head;
+
+    while (lastEl.next) {
+      penultEl = lastEl;
+      lastEl = lastEl.next;
+    }
+
+    if (!penultEl) {
+      this.head = undefined;
+    } else {
+      penultEl.next = undefined;
+    }
+  }
+
+  reverse() {
+    if (!this.head) {
+      return;
+    }
+
+    let current;
+    let previous;
+
+    while (this.head) {
+      current = this.head;
+      this.shift();
+      current.next = previous;
+      previous = current;
+    }
+
+    this.head = previous;
+  }
+
+  toArray() {
+    const array = [];
+
+    while (this.head) {
+      array.push(this.head.value);
+      this.shift();
```

Comment:
I am not sure of the javascript implementation, but this operation is destructive. Since the head is moving forward, once we do `list.toArray()`, there is no way to get back the linked list - which should not be the case. Correct me if I am wrong.

Reply:
You're right - and I should port this patch to the JavaScript solution too! Thanks for catching!